### PR TITLE
added solution dao-8

### PR DESCRIPTION
**Pull request by IgorBochkor**
added solution dao-8

### CODE REVIEW

File: src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        Long driverId = Long.valueOf(req.getParameter("driverId"));
+        Long carId = Long.valueOf(req.getParameter("carId"));
+        Car car = carService.get(carId);
+        Driver driver = driverService.get(driverId);
+        carService.addDriverToCar(driver, car);
```
Comment (yurwing) [inline src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java:34]: here also check input data

File: src/main/java/com/igorbochkor/controllers/GetAllDriversController.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+public class GetAllDriversController extends HttpServlet {
+    private static final Injector injector = Injector.getInstance("com.igorbochkor");
+    private DriverService driverService
+            = (DriverService) injector.getInstance(DriverService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+
```
Comment (yurwing) [inline src/main/java/com/igorbochkor/controllers/GetAllDriversController.java]: empty line

**Review by yurwing — CHANGES_REQUESTED** [review]

File: src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java (hunk 1 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java:34]: done!

File: src/main/java/com/igorbochkor/controllers/GetAllDriversController.java (hunk 2 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/GetAllDriversController.java]: done

**Review by yurwing — APPROVED** [review]

File: src/main/java/com/igorbochkor/Main.java [hunk 3]
```diff
@@ -52,6 +46,6 @@ public static void main(String[] args) {
         carService.addDriverToCar(driverTwo, carTwo);
         System.out.println(carService.getAll());
         System.out.println(carService.getAllByDriver(driverTwo.getId()));
-        carService.delete(car.getId());
+        carService.delete(car.getId());*/
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/Main.java]: can we remove the commented code? probably we can remove the Main class

File: src/main/java/com/igorbochkor/controllers/AddCarController.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/car/add.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String model = req.getParameter("model");
+        String manufacturerIdString = req.getParameter("manufacturerId");
+        if (!model.isEmpty() && manufacturerIdString.matches("\\d+")) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddCarController.java]: let's remove these checks - we can do this validation on the JSP

File: src/main/java/com/igorbochkor/controllers/AddCarController.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            Long manufacturerId = Long.valueOf(manufacturerIdString);
+            resp.sendRedirect(req.getContextPath() + "/");
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddCarController.java]: this should be the last line - after we create the car

File: src/main/webapp/WEB-INF/views/car/add.jsp [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    <title>Create car</title>
+</head>
+<body>
+<h1>Please provide information about car: model and manufacturerId</h1>
+
+<h4 style="color: brown">${message}</h4>
+
+<form method="post" action="${pageContext.request.contextPath}/car/add">
+    Please provide model: <input type="text" name="model">
+    Please provide manufacturerId: <input type="text" name="manufacturerId">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/add.jsp]:
```suggestion
<form method="post" action="${pageContext.request.contextPath}/cars/add">
    Please provide model: <input type="text" name="model" required>
    Please provide manufacturerId: <input type="number" name="manufacturerId" required>
```

File: src/main/java/com/igorbochkor/controllers/AddDriverController.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String licenceNumber = req.getParameter("licenceNumber");
+
+        if (!name.isEmpty() && !licenceNumber.isEmpty()) {
+            resp.sendRedirect(req.getContextPath() + "/");
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddDriverController.java]: same comments here

File: src/main/webapp/WEB-INF/views/drivers/add.jsp [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+<head>
+    <title>Create driver</title>
+</head>
+<body>
+<h1>Please provide information about driver: name and license number</h1>
+
+<h4 style="color: brown">${message}</h4>
+<form method="post" action="${pageContext.request.contextPath}/drivers/add">
+    Please provide name: <input type="text" name="name">
+    Please provide license number: <input type="text" name="licenceNumber">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/drivers/add.jsp:12]: same

File: src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/car/addDriver.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String driverIdString = req.getParameter("driverId");
+        String carIdString = req.getParameter("carId");
+        if (driverIdString.matches("\\d+") && carIdString.matches("\\d+")) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java]: same here

File: src/main/webapp/WEB-INF/views/car/addDriver.jsp [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    <title>Add driver for car</title>
+</head>
+<body>
+<h1>Please provide information adding driver for car: Driver id and Car id</h1>
+
+<h4 style="color: brown">${message}</h4>
+
+<form method="post" action="${pageContext.request.contextPath}/car/addDriver">
+    Please provide Driver id: <input type="text" name="driverId">
+    Please provide Car id: <input type="text" name="carId">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/addDriver.jsp]:
```suggestion
<form method="post" action="${pageContext.request.contextPath}/car/addDriver">
    Please provide Driver id: <input type="number" name="driverId" required>
    Please provide Car id: <input type="number" name="carId" required>
```

File: src/main/java/com/igorbochkor/controllers/AddManufacturerController.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/manufacturer/add.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String country = req.getParameter("country");
+        if (!name.isEmpty() && !country.isEmpty()) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddManufacturerController.java]: same

File: src/main/java/com/igorbochkor/controllers/AddManufacturerController.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+            resp.sendRedirect(req.getContextPath() + "/");
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/AddManufacturerController.java]: make this the last line

File: src/main/java/com/igorbochkor/controllers/GetAllDriversController.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    private static final Injector injector = Injector.getInstance("com.igorbochkor");
+    private DriverService driverService
+            = (DriverService) injector.getInstance(DriverService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        List<Driver> allDrivers = driverService.getAll();
+        req.setAttribute("drivers", allDrivers);
+        req.getRequestDispatcher("/WEB-INF/views/drivers/get/all.jsp").forward(req, resp);
```
Comment (KseniiaMakarova) [inline src/main/java/com/igorbochkor/controllers/GetAllDriversController.java]:
```suggestion
        req.getRequestDispatcher("/WEB-INF/views/drivers/all.jsp").forward(req, resp);
```

File: src/main/webapp/web.xml [hunk 14]
```diff
@@ -11,6 +11,56 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>allDrivers</servlet-name>
+        <servlet-class>com.igorbochkor.controllers.GetAllDriversController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>allDrivers</servlet-name>
+        <url-pattern>/drivers/get/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/drivers/</url-pattern>
```

File: src/main/webapp/web.xml [hunk 15]
```diff
@@ -11,6 +11,56 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addManufacturer</servlet-name>
+        <servlet-class>com.igorbochkor.controllers.AddManufacturerController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addManufacturer</servlet-name>
+        <url-pattern>/manufacturer/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/manufacturers/add</url-pattern>
```

File: src/main/webapp/web.xml [hunk 16]
```diff
@@ -11,6 +11,56 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addCar</servlet-name>
+        <servlet-class>com.igorbochkor.controllers.AddCarController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addCar</servlet-name>
+        <url-pattern>/car/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/cars/add</url-pattern>
```

File: src/main/webapp/web.xml [hunk 17]
```diff
@@ -11,6 +11,56 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>com.igorbochkor.controllers.AddDriverToCarController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/addDriver</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/cars/drivers/add</url-pattern>
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/igorbochkor/Main.java (hunk 3 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/Main.java]: removed

File: src/main/java/com/igorbochkor/controllers/AddCarController.java (hunk 4 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddCarController.java]: done

File: src/main/java/com/igorbochkor/controllers/AddCarController.java (hunk 5 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddCarController.java]: done

File: src/main/webapp/WEB-INF/views/car/add.jsp (hunk 6 above)
Comment (IgorBochkor) [reply src/main/webapp/WEB-INF/views/car/add.jsp]: added

File: src/main/java/com/igorbochkor/controllers/AddDriverController.java (hunk 7 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddDriverController.java]: done

File: src/main/webapp/WEB-INF/views/drivers/add.jsp (hunk 8 above)
Comment (IgorBochkor) [reply src/main/webapp/WEB-INF/views/drivers/add.jsp:12]: done

File: src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java (hunk 9 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddDriverToCarController.java]: done

File: src/main/webapp/WEB-INF/views/car/addDriver.jsp (hunk 10 above)
Comment (IgorBochkor) [reply src/main/webapp/WEB-INF/views/car/addDriver.jsp]: added

File: src/main/java/com/igorbochkor/controllers/AddManufacturerController.java (hunk 11 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddManufacturerController.java]: done

File: src/main/java/com/igorbochkor/controllers/AddManufacturerController.java (hunk 12 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/AddManufacturerController.java]: done

File: src/main/java/com/igorbochkor/controllers/GetAllDriversController.java (hunk 13 above)
Comment (IgorBochkor) [reply src/main/java/com/igorbochkor/controllers/GetAllDriversController.java]: done

File: src/main/webapp/web.xml (hunk 14 above)
Comment (IgorBochkor) [reply src/main/webapp/web.xml]: done

File: src/main/webapp/web.xml (hunk 15 above)
Comment (IgorBochkor) [reply src/main/webapp/web.xml]: done

File: src/main/webapp/web.xml (hunk 16 above)
Comment (IgorBochkor) [reply src/main/webapp/web.xml]: done

File: src/main/webapp/web.xml (hunk 17 above)
Comment (IgorBochkor) [reply src/main/webapp/web.xml]: done

File: src/main/webapp/web.xml [hunk 18]
```diff
@@ -11,6 +11,56 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>com.igorbochkor.controllers.AddDriverToCarController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/driver/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:64]:
```suggestion
        <url-pattern>/cars/drivers/add</url-pattern>
```

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**